course-content/31_exploring_data_structures.qmd

```diff
@@ -0,0 +1,447 @@
+:::
+:::{.callout-note icon="false" collapse="true"}
+
+### Solution to Challenge 2
+
+First you would like to create a new object to hold this narrowed down version of the dataset, then use some method like `.print()` or `.head()` to see some of the resulting data. 
+
+
+```{python}
+challenge2_df = df[(df['year'] >= 1992) & (df['country'] == 'Canada')]
+print(challenge2_df)
+```
+:::
+
+
+You will find that it is very common within python/pandas to see the `.iloc()` method.
+
+The `.iloc()` method identifies rows based on their index value (that column to the left of the dataframe when you call the <code>.head()</code> function, for example). It can be used to look at individual rows of data by their original index value.
+
+```{python}
+df.iloc[251]
+```
+
+Or it can be used to call multiple rows in a range:
+
+```{python}
+df.iloc[0:5]
+```
+
+## Sorting rows
+
+So now you've mastered how to select rows and columns you want, congratulations! Instead of hunting and pecking for insights, one way to quickly make some sense of the data is to sort it - something you probably do in Excel all the time. 
+
+In the challenge above we created a dataframe "challenge2_df" with just the one country in it and from 1997. So now let's sort the data by year, and then see what can spot any insights. To do this, we use the `.sort_values()` method.
+
+Let's go fetch and look at the dataframe, and use this method to sort by year, going from more recent to less recent:
+
+```{python}
+challenge2_df.sort_values('year', ascending=False)
+```
+
+## Putting multiple methods together (chaining)
+
+You can also put multiple methods in the same line of code, as long as you separate each of the elements. In pandas, this is called "method chaining",  as you are essentially of creating a chain of actions to take place. 
+
+In the example below, the chain consists of two links:
+
+1. Select just the rows where the country is equal to Canada AND where the year is 1997 or greater
+2. Take these rows and sort them by year, going from high to low values, top to bottom
+
+```{python}
+#| scrolled: true
+chained_df = df[(df['country'] == 'Canada') & (df['year'] >= 1997)].sort_values('year', ascending = False)
+chained_df.head(3)
+```
+
+Another way of chaining is to put each of the functions on multiple lines.  In the example below, each line of code modifies the dataframe to acheive the same result as above.
+
+```{python}
+chained2_df = df[(df['country'] == 'Canada')]
+chained2_df = chained2_df[(chained2_df['year'] >= 1997)]
+chained2_df = chained2_df.sort_values('year', ascending=False)
+chained2_df.head(3)
+```
+
+This latter approach is very common.  Each respective line to the right of the `=` contains the code that modifies the object assigned to the left of it. Note that you can use a combination of chaining and piping.  
```

Comment:
```suggestion
This latter approach is very common, as it can be easier to read the code and interpret exactly what is happening. Each respective line to the right of the `=` contains the code that modifies the object assigned to the left of it. Note that you can use a combination of chaining and piping.  
```

You used to have a section that explained chaining but on multiple lines - I would keep this somewhere in the chaining section:
```{python}
chained3_df = df[(df['country'] == 'Canada') & (df['year'] >= 1997])] \
   .sort_values('year', ascending = False)
```